optimize Linode SD by polling for event changes during refresh

tjhop opened the pull request:
Most accounts are fairly "static", in the sense that they're not cycling
through instances constantly. So rather than do a full refresh every
interval and potentially make several behind-the-scenes paginated API
calls, this will now poll the `/account/events/` endpoint every minute
with a list of events that we care about. If a matching event is found,
we then do a full refresh.

Co-authored-by: William Smith <wsmith@linode.com>
Signed-off-by: TJ Hoplock <t.hoplock@gmail.com>
Signed-off-by: William Smith <wsmith@linode.com>

roidelapluie commented:
The change of permissions of the token is a bit annoying for users who would upgrade. Would there be a possibility to have a fallback if the token does not have the permission for listing events?

Also, what if the list of event changes in the future, or we miss one? What is the stability here? Would it be better to work with an exclude list instead of an include list?

It also seems like d.lastRefreshTimestamp is set too late and if listing 10 pages takes 10s, we would miss all the updates during the 10s now.

Dorthu commented:
> The change of permissions of the token is a bit annoying for users who would upgrade. Would there be a possibility to have a fallback if the token does not have the permission for listing events?

I suppose it could fall back to reloading everything on every refresh if it sees a 401 response when attempting to fetch events the first time.

> Also, what if the list of event changes in the future, or we miss one? What is the stability here? Would it be better to work with an exclude list instead of an include list?

That's a fair concern.  We could remove all the filters on event action and this would still be a big improvement; events are only created when there are changes being made to the account, and I believe it will be more common for there to be no events most refreshes.

> It also seems like d.lastRefreshTimestamp is set too late and if listing 10 pages takes 10s, we would miss all the updates during the 10s now.

Good catch 👍

tjhop commented:
I've rebased to current and added in a way to fallback to full refreshes if event polling isn't possible. Lemme know if you have any other feedback! 👍

roidelapluie commented:
Is there a bug in the linode API? It seems that if also need account:read_only to read the events. events_readonly is not enough.

I have run a simple test and it seems that accepting this pull request would cause a regression: the nodes `__meta_linode_status` label is no longer updated when the node is shutdown manually (and lassie is disabled). I do not have any idea about how to fix this at this moment. What about we still trigger a full refresh every 10*refresh_interval or so?

Dorthu commented:
It turns out that this indeed is a bug in the API; it's being fixed in the next release scheduled for 7/12.  After that point, only the `events:read_only` scope will be necessary to list Events.  We should hold off on this proposed change until the fix is released.

tjhop commented:
Hi @roidelapluie, the `events` token scope has been fixed. 👍 

> I have run a simple test and it seems that accepting this pull request would cause a regression: the nodes __meta_linode_status label is no longer updated when the node is shutdown manually (and lassie is disabled). I do not have any idea about how to fix this at this moment. What about we still trigger a full refresh every 10*refresh_interval or so?

Polling the `/events` endpoint will only return events that the Linode API is aware of, so you're correct -- if you manually shut down an instance, the API won't have an event for it. I added in a counter to trigger a full refresh after 10 requests to the `/events` endpoint without changes (since we now trigger a refresh on _any_ event detected). This should catch the edge case you mentioned and ensure that we periodically refresh regardless.

Lemme know if you have any other feedback!

roidelapluie commented on discovery/linode/linode.go:
```diff
@@ -143,6 +156,50 @@ func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+
+		// Check to see if there have been any events. If so, refresh our data.
+		opts := linodego.NewListOptions(1, fmt.Sprintf(filterTemplate, d.lastRefreshTimestamp.Format("2006-01-02T15:04:05")))
+		events, err := d.client.ListEvents(ctx, opts)
```
This seems like it will only fetch the events in the last second?

tjhop replied:
Addressed in f61624850c064dc0b7933fcb70c50adeaf383462

roidelapluie commented:
I would take the date before the refresh:

Pseudocode

```
nextRefreshdate = time.Now
err = ListEvents(d.lastRefreshDate)
if err == nil {
d.lastRefreshDate = nextRefreshdate
}
```

tjhop commented:
Good catch, and my thoughts as well. I'll make the change 👍

tjhop commented:
Hey @roidelapluie, just following up to see if you have any further thoughts or feedback on the proposed changes. Please lemme know if I can help 👍

roidelapluie commented on discovery/linode/linode.go:
```diff
@@ -185,6 +184,8 @@ func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
 			}
 		}
+
+		d.lastRefreshTimestamp = ts
```
I would move this after we know that the refresh went well, e.g. L199

roidelapluie commented:
thanks!

tjhop commented:
Thanks for all the help @roidelapluie!
